src/sysupdate/sysupdated.c

```diff
@@ -0,0 +1,1857 @@
+                t->id = strdup("host"); /* This is what appears in the object path */
+        else
+                t->id = strjoin(target_class_to_string(class), ":", name);
+        if (!t->id)
+                return -ENOMEM;
+
+        r = hashmap_ensure_put(&m->targets, &target_hash_ops, t->id, t);
+        if (r < 0)
+                return r;
+
+        *ret = TAKE_PTR(t);
+        return 0;
+}
+
+static int sysupdate_run_simple(sd_json_variant **ret, ...) {
+        _cleanup_close_pair_ int pipe[2] = EBADF_PAIR;
+        _cleanup_(pidref_done_sigkill_wait) PidRef pid = PIDREF_NULL;
+        _cleanup_fclose_ FILE *f = NULL;
+        _cleanup_(sd_json_variant_unrefp) sd_json_variant *v = NULL;
+        int r;
+
+        r = pipe2(pipe, O_CLOEXEC);
+        if (r < 0)
+                return r;
+
+        r = pidref_safe_fork_full("(sd-sysupdate)",
+                                  (int[]) { -EBADF, pipe[1], STDERR_FILENO },
+                                  NULL, 0,
+                                  FORK_RESET_SIGNALS|FORK_CLOSE_ALL_FDS|FORK_DEATHSIG_SIGTERM|
+                                  FORK_REARRANGE_STDIO|FORK_LOG|FORK_REOPEN_LOG,
+                                  &pid);
+        if (r < 0)
+                return r;
+        if (r == 0) {
+                /* Child */
+                va_list ap;
+                char *arg;
+                _cleanup_strv_free_ char **args = NULL;
+
+                if (strv_extend(&args, "systemd-sysupdate") < 0) {
+                        log_oom();
+                        _exit(EXIT_FAILURE);
+                }
+
+                if (strv_extend(&args, "--json=short") < 0) {
+                        log_oom();
+                        _exit(EXIT_FAILURE);
+                }
+
+                va_start(ap, ret);
+                while ((arg = va_arg(ap, char*)))
+                        if (strv_extend(&args, arg) < 0) {
+                                log_oom();
+                                _exit(EXIT_FAILURE);
+                        }
+                va_end(ap);
+
+                if (DEBUG_LOGGING) {
+                        _cleanup_free_ char *s = NULL;
+
+                        s = quote_command_line((char**) args, SHELL_ESCAPE_EMPTY);
+                        if (!s) {
+                                log_oom();
+                                _exit(EXIT_FAILURE);
+                        }
+
+                        log_debug("Spawning sysupdate: %s", s);
+                }
+
+                r = invoke_callout_binary(sysupdate_binary_path(), args);
+                log_error_errno(r, "Failed to execute systemd-sysupdate: %m");
+                _exit(EXIT_FAILURE);
+        }
+
+        pipe[1] = safe_close(pipe[1]);
+        f = take_fdopen(&pipe[0], "r");
+        if (!f)
+                return -errno;
+
+        r = sd_json_parse_file(f, "stdout", 0, &v, NULL, NULL);
+        if (r < 0)
+                return log_error_errno(r, "Failed to parse JSON: %m");
+
```

Comment:
Yeah I wrote this specifically for only quick operations and to avoid callback hell... Maybe it makes a bit less sense now that things have been reorganized a little?

I suppose it all depends on your definition of "quick". Is looking up the installed versions on the host system quick? IMO it is, but you might disagree. Enumerating components (just walking dirs in /usr/lib`)?

Reply:
i guess we can leave it at this for now, maybe address later if this becomes a problem.

i just want that you thought about it, that's all